### PR TITLE
refactor-mesh-lookup - Use EnTT resource management for game assets

### PR DESCRIPTION
**Pull request by raffclar**

### CODE REVIEW

File: src/3D/L3DMesh.cpp [hunk 1]
```diff
@@ -131,8 +131,10 @@ void L3DMesh::LoadFromBuffer(const std::vector<uint8_t>& data)
 	catch (std::runtime_error& err)
 	{
 		SPDLOG_LOGGER_ERROR(spdlog::get("game"), "Failed to open l3d mesh from buffer: {}", err.what());
-		return;
+		return false;
 	}
 
 	Load(l3d);
+
+	return true;
```
Comment (raffclar) [inline src/3D/L3DMesh.cpp:139]: Remove space above.

File: src/3D/MeshPack.cpp [hunk 2]
```diff
@@ -54,28 +54,12 @@ bool MeshPack::LoadFromFile(const fs::path& path)
...
-	// TODO(raffclar): Handle non-mesh pack IDs via a new mechanism
-	if (static_cast<int>(id) == ecs::components::Hand::meshId)
-	{
-		return Game::instance()->GetHandModel();
-	}
-	else
-	{
-		return *_meshes.at(static_cast<size_t>(id));
-	}
+	return *_meshes.at(static_cast<size_t>(id));
```
Comment (raffclar) [inline src/3D/MeshPack.cpp:57]: Remove MeshPack (?).

File: src/3D/MeshPack.cpp [hunk 3]
```diff
@@ -54,28 +54,12 @@ bool MeshPack::LoadFromFile(const fs::path& path)
...
```
Comment (raffclar) [inline src/3D/MeshPack.cpp:62]: Remove MeshPack (?).

File: src/ECS/Systems/RenderingSystem.cpp [hunk 4]
```diff
@@ -102,9 +103,9 @@ void RenderingSystem::PrepareDrawUploadUniforms(bool drawBoundingBox)
 		    _renderContext.instanceUniforms[idx] = modelMatrix;
 		    if (drawBoundingBox)
 		    {
-			    const L3DMesh& l3dMesh = Game::instance()->GetMeshPack().GetMesh(mesh.id);
-			    auto submeshId = (l3dMesh.GetNumSubMeshes() + mesh.bbSubmeshId) % l3dMesh.GetNumSubMeshes();
-			    auto box = l3dMesh.GetSubMeshes()[submeshId]->GetBoundingBox();
+		    	auto l3dMesh = entt::service_locator<resources::ResourcesInterface>::ref().GetMeshes().Handle(mesh.id);
```
Comment (raffclar) [inline src/ECS/Systems/RenderingSystem.cpp:106]: Broken indent

File: src/Game.cpp [hunk 5]
```diff
@@ -337,8 +336,43 @@ bool Game::Update()
 	return _config.numFramesToSimulate == 0 || _frameCount < _config.numFramesToSimulate;
 }
 
+
+using namespace entt::literals;
```
Comment (raffclar) [inline src/Game.cpp:340]: Remove this. Literals aren't being used.

File: src/Locator.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+namespace resources
+{
+using MeshManager = ResourceManager<resources::L3DLoader>;
+using TextureManager = ResourceManager<resources::Texture2DLoader>;
+using AnimationManager = ResourceManager<resources::L3DAnimLoader>;
+
+class ResourcesInterface
+{
+public:
+	ResourcesInterface() {}
```
Comment (raffclar) [inline src/Locator.h:29]: Not needed

File: src/Locator.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	virtual ~ResourcesInterface() {}
```
Comment (raffclar) [inline src/Locator.h:30]: Not needed

File: src/Locator.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	resources::TextureManager _textures;
+	resources::AnimationManager _animations;
+};
+} // namespace resources
+
+struct Locator
+{
+	using resources = entt::service_locator<resources::ResourcesInterface>;
+};
+} // namespace openblack
```
Comment (raffclar) [inline src/Locator.h:55]: No EoL

File: src/Resources/ResourceManager.h [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		_resourceCache.each(func);
+	}
+
+	[[nodiscard]] entt::resource_cache<ResourceType>& GetCache() const { return _resourceCache; }
+
+	[[nodiscard]] decltype(auto) Size() const { return _resourceCache.size(); }
+private:
+	entt::resource_cache<ResourceType> _resourceCache;
+};
+} // openblack::resources
```
Comment (raffclar) [inline src/Resources/ResourceManager.h:71]: No EoL

File: src/3D/MeshPack.cpp (hunk 3 above)
Comment (raffclar) [reply src/3D/MeshPack.cpp:62]: Removed.

**Comment by raffclar** [comment]
Fixing the CI failure. Looks like a change was made to fmtlib.

File: src/Resources/Loaders.h [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+template <typename Loader, typename Resource>
+struct BaseLoader: entt::resource_loader<Loader, Resource>
+{
+public:
+	using ResourceType = Resource;
+};
+
+struct L3DLoader final: BaseLoader<L3DLoader, L3DMesh>
+{
+	[[nodiscard]] std::shared_ptr<L3DMesh> load(const std::string& debugName, const std::vector<uint8_t>& data) const
```
Comment (bwrsandman) [inline src/Resources/Loaders.h:32]: Not of fan of using shared pointers. The reason for this is that it makes ownership impossible. The resource just stays around as long as there's a reference.
Isn't the point of using resource management that entt owns the resource?

File: src/Resources/Loaders.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+		if (!mesh->LoadFromFile(path))
+		{
+			throw std::runtime_error("Unable to load mesh");
+		}
+
+		return mesh;
+	}
+};
+
+struct Texture2DLoader final: BaseLoader<Texture2DLoader, graphics::Texture2D>
```
Comment (bwrsandman) [inline src/Resources/Loaders.h:55]: Could you add a loader for `.raw` textures from `filesystem.TexturePath()`

File: src/Resources/Resources.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+	resources::MeshManager& GetMeshes() override { return _meshes; }
+	resources::TextureManager& GetTextures() override { return _textures; }
+	resources::AnimationManager& GetAnimations() override { return _animations; }
+
+private:
+	resources::MeshManager _meshes;
+	resources::TextureManager _textures;
+	resources::AnimationManager _animations;
+};
+} // namespace openblack::resources
```
Comment (bwrsandman) [inline src/Resources/Resources.h:42]: EOL

File: src/Resources/Loaders.h (hunk 10 above)
Comment (raffclar) [reply src/Resources/Loaders.h:32]: entt resource management relies on shared ptrs
https://github.com/skypjack/entt/blob/master/docs/md/resource.md

Comment (bwrsandman) [reply src/Resources/Loaders.h:32]: Then let's make our accesses with weak_ptr

Comment (raffclar) [reply src/Resources/Loaders.h]: Changed to use entt::resource_handle as the return type. It's basically an alias for shared_ptr but required by the interface spec.

File: src/Resources/Loaders.h (hunk 11 above)
Comment (raffclar) [reply src/Resources/Loaders.h:55]: Yeah I can do that. Should I do this in a new PR? This is getting big.

**Comment by raffclar** [comment]
[7](https://github.com/openblack/openblack/pull/341/commits/7debda991a40313428679a79a38a13d6e75df1c1) will be dropped once approved

File: src/3D/AllMeshes.cpp [hunk 13]
```diff
@@ -0,0 +1,643 @@
...
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "AllMeshes.h"
+
+namespace openblack
+{
+
+std::array<std::string, static_cast<uint16_t>(MeshId::_COUNT)> MeshNames = {
```
Comment (bwrsandman) [inline src/3D/AllMeshes.cpp:15]: In general +1 for moving this out of the header.
Could you make this a `constexpr` or `constinit` to make it threadsafe
Second, if there is any indication that this array can come from a resource file from the game, please make a note of it and write your code in function of the fact that in the future we want to load those strings from data files

File: src/ECS/Archetypes/FeatureArchetype.cpp [hunk 14]
```diff
@@ -38,20 +39,21 @@ entt::entity FeatureArchetype::Create(const glm::vec3& position, FeatureInfo typ
 	const auto [point, radius] = GetFixedObstacleBoundingCircle(info.meshId, transform);
 	registry.Assign<Fixed>(entity, point, radius);
 	registry.Assign<Feature>(entity, type);
-	const auto& mesh = registry.Assign<Mesh>(entity, info.meshId, static_cast<int8_t>(0), static_cast<int8_t>(1));
+	const auto resourceId = MeshIdToResourceId(info.meshId);
+	const auto& mesh = registry.Assign<Mesh>(entity, resourceId, static_cast<int8_t>(0), static_cast<int8_t>(1));
```
Comment (bwrsandman) [inline src/ECS/Archetypes/FeatureArchetype.cpp:43]: You can remove this line

File: src/ECS/Archetypes/Utils.cpp [hunk 15]
```diff
@@ -14,17 +14,20 @@
...
 
 using namespace openblack;
 using namespace openblack::ecs::components;
 
 std::pair<glm::vec2, float> openblack::ecs::archetypes::GetFixedObstacleBoundingCircle(MeshId meshId,
                                                                                        const Transform& transform)
 {
-	const auto& bb = Game::instance()->GetMeshPack().GetMesh(meshId).GetBoundingBox();
+	auto resourceId = MeshIdToResourceId(meshId);
+	auto l3dMesh = Locator::resources::ref().GetMeshes().Handle(resourceId);
```
Comment (bwrsandman) [inline src/ECS/Archetypes/Utils.cpp:29]: If they can be const please mark them as const by default

File: src/Game.cpp [hunk 16]
```diff
@@ -386,6 +384,37 @@ bool Game::Update()
...
+	auto& meshManager = resources.GetMeshes();
+	auto& textureManager = resources.GetTextures();
+	auto& animationManager = resources.GetAnimations();
+
+	meshManager.Load("hand", "hand", (_fileSystem->CreatureMeshPath() / "Hand_Boned_Base2.l3d").u8string());
+	meshManager.Load("testModel", "test_model", (_fileSystem->MiscPath() / "coffre.l3d").u8string());
+	pack::PackFile pack;
+	pack.Open(Game::instance()->GetFileSystem().FindPath(_fileSystem->DataPath() / "AllMeshes.g3d").u8string());
+	auto& meshes = pack.GetMeshes();
+	for (auto i = 0; i < meshes.size(); i++)
```
Comment (bwrsandman) [inline src/Game.cpp:398]: Getting some warnings here about the type of `i = 0` being int instead of size_t

File: src/Game.cpp [hunk 17]
```diff
@@ -386,6 +384,37 @@ bool Game::Update()
...
+	for (auto const& [name, g3dTexture] : textures)
+	{
+		textureManager.Load(g3dTexture.header.id, name, g3dTexture);
+	}
+
+	animationManager.Load("testAnimation", (_fileSystem->MiscPath() / "coffre.anm").u8string());
+	pack::PackFile animationPack;
+	animationPack.Open(Game::instance()->GetFileSystem().FindPath(_fileSystem->DataPath() / "AllAnims.anm").u8string());
+	auto& animations = animationPack.GetAnimations();
+	for (auto i = 0; i < animations.size(); i++)
```
Comment (bwrsandman) [inline src/Game.cpp:413]: here too

File: src/Game.cpp [hunk 18]
```diff
@@ -386,6 +384,37 @@ bool Game::Update()
...
+	auto& animationManager = resources.GetAnimations();
+
+	meshManager.Load("hand", "hand", (_fileSystem->CreatureMeshPath() / "Hand_Boned_Base2.l3d").u8string());
+	meshManager.Load("testModel", "test_model", (_fileSystem->MiscPath() / "coffre.l3d").u8string());
+	pack::PackFile pack;
+	pack.Open(Game::instance()->GetFileSystem().FindPath(_fileSystem->DataPath() / "AllMeshes.g3d").u8string());
+	auto& meshes = pack.GetMeshes();
+	for (auto i = 0; i < meshes.size(); i++)
+	{
+		auto m = meshManager.Load(i, MeshNames[i], meshes[i]);
```
Comment (bwrsandman) [inline src/Game.cpp:400]: Is `auto m` here to prevent a warning about disposing `m`?

File: src/Resources/Loaders.h (hunk 11 above)
Comment (bwrsandman) [reply src/Resources/Loaders.h:55]: Yes, could you make an issue for that?

File: src/Resources/Loaders.h [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+		auto mesh = std::make_shared<L3DMesh>(debugName);
+		if (!mesh->LoadFromBuffer(data))
+		{
+			throw std::runtime_error("Unable to load mesh");
+		}
+
+		return mesh;
+	}
+
+	[[nodiscard]] entt::resource_handle<L3DMesh> load(const std::string& debugName, const std::string& path) const
```
Comment (bwrsandman) [inline src/Resources/Loaders.h:43]: Are you sure you want `nodiscard` here?

File: src/Resources/MeshId.h [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+#include "3D/AllMeshes.h"
+
+template <>
+struct fmt::formatter<openblack::MeshId>: fmt::formatter<uint32_t>
+{
+	auto format(openblack::MeshId c, format_context& ctx) { return formatter<uint32_t>::format(static_cast<uint32_t>(c), ctx); }
+};
+
+static entt::id_type MeshIdToResourceId(openblack::MeshId id)
```
Comment (bwrsandman) [inline src/Resources/MeshId.h:20]: Warnings about this function

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Some changes needed but over all I would say it's very well done.
Great work!

File: src/Resources/Loaders.h (hunk 11 above)
Comment (raffclar) [reply src/Resources/Loaders.h:55]: Added #369

File: src/3D/AllMeshes.cpp (hunk 13 above)
Comment (raffclar) [reply src/3D/AllMeshes.cpp:15]: constexpr std::string support and constinit are both added in c++20

File: src/Game.cpp (hunk 18 above)
Comment (raffclar) [reply src/Game.cpp:400]: Nicely spotted. Just debugging leftover

File: src/Resources/Loaders.h (hunk 19 above)
Comment (raffclar) [reply src/Resources/Loaders.h:43]: Clang tidy tells me bad things if I remove them. Generally shouldn't be an issue as EnTT will be the caller of these `load` functions.

File: src/Resources/MeshId.h (hunk 20 above)
Comment (raffclar) [reply src/Resources/MeshId.h:20]: image

Comment (raffclar) [reply src/Resources/MeshId.h:20]: So it's used but yeah this warning exists:
 Warning: /home/runner/work/openblack/openblack/src/Resources/MeshId.h:20:22: warning: unused function 'MeshIdToResourceId' [-Wunused-function]
  static entt::id_type MeshIdToResourceId(openblack::MeshId id)

Comment (bwrsandman) [reply src/Resources/MeshId.h:20]: Try putting it in a namespace see if that helps

File: src/ECS/Archetypes/FeatureArchetype.cpp (hunk 14 above)
Comment (raffclar) [reply src/ECS/Archetypes/FeatureArchetype.cpp:43]: What do you mean?

Comment (raffclar) [reply src/ECS/Archetypes/FeatureArchetype.cpp]: Removed `mesh` assignment

File: src/Resources/Loaders.h (hunk 19 above)
Comment (raffclar) [reply src/Resources/Loaders.h]: Is this ok?

**Comment by bwrsandman** [comment]
Will approve after some of the header code is moved to implementation files.

**Comment by raffclar** [comment]
> Will approve after some of the header code is moved to implementation files.

Moved the loader code to its own unit but I couldn't do much with the `ResourceManager` class. It doesn't have any methods tha can be refactored.

**Review by bwrsandman — APPROVED** [review]

**Comment by raffclar** [comment]
Rebasing

File: src/Resources/Loaders.cpp [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "Resources/Loaders.h"
+
+namespace openblack::resources
```
Comment (bwrsandman) [inline src/Resources/Loaders.cpp:12]: You can just have `using namespace openblack::resources`. There's nothing new being defined

Comment (raffclar) [reply src/Resources/Loaders.cpp]: Fixed

**Pull request merged**